blog/tech-talk-window-resize-behavior.md

```diff
@@ -0,0 +1,290 @@
+2. While resizing, switch from an `IDXGISwapChain1` to a DirectComposition surface which synchronizes updates with `IDCompositionDevice::Commit`. This also fixes the artifacts.
+
+We went with the first option because it leads to faster resizes than the second option.
+
+I landed a [patch](https://crrev.com/c/7115438) in Chromium that implemented that first solution.
+
+I also submitted two other patches in preparation for the main patch:
+
+1. The [first one](https://crrev.com/c/7129658) fixed a bug in existing code that would make CI fail in combination with the main patch. It also made launching Electron apps and Chrome a tiny bit faster.
+2. The [second one](https://crrev.com/c/7210913) was split off to make code review easier on the main patch.
+
+## Second bug
+
+In addition to this first bug, there was a second bug which also led to stale pixels.
+
+Here's what was going on there:
+
+When the user resizes the window, Chromium needs to redraw the contents of the window for the new window size. This takes some time. The new frame isn't ready immediately.
+
+Here's a sequence of frames that demonstrates this:
+
+![First frame of the Chromium resize sequence](/assets/img/blog/tech-talk-window-resize-behavior/chromium-resize-sequence-1.png)
+![Second frame of the Chromium resize sequence](/assets/img/blog/tech-talk-window-resize-behavior/chromium-resize-sequence-2.png)
+![Third frame of the Chromium resize sequence](/assets/img/blog/tech-talk-window-resize-behavior/chromium-resize-sequence-3.png)
+![Fourth frame of the Chromium resize sequence](/assets/img/blog/tech-talk-window-resize-behavior/chromium-resize-sequence-4.png)
+![Fifth frame of the Chromium resize sequence](/assets/img/blog/tech-talk-window-resize-behavior/chromium-resize-sequence-5.png)
+
+At a certain time during the resize, Windows tells us: "The window is 1,000 pixels wide." But the frames that the browser compositor produces are lagging behind. The last frame that we have painted might be 600 pixels wide.
+
+Historically, Chromium used to skip frames where the width of the window did not match the width of the frame that it last painted. It would decide to just not update the window.
+
+However, that would often lead to the window contents not being updated at all until the resize operation was finished.
+
+So [in 2015](https://crrev.com/1426453006) someone decided: "Why not show these frames? They might not match the window size perfectly, but at least we can show _something_."
+
+It would lead to gutter, but at that time the gutter was black. So that was better than the previous implementation.
+
+Now, 10 years later with DirectComposition, that gutter was often filled with stale pixels.
+
+Let's look at what was happening there:
+
+Every frame consists of multiple render passes. The render passes represent the various things that should be drawn on screen. From complicated bitmaps to rectangles filled with solid colors.
```

Comment:
```suggestion
Every frame consists of multiple render passes. These render passes represent the various things that should be drawn on screen, from complicated bitmaps to rectangles filled with solid colors.
```

Reply:
Thanks! Changed "The" -> "These". With the "from" part, I'm bending the rules of grammar a bit. But I find it much more readable to separate it with a dot than with a comma.